File: 1039-find-the-town-judge/1039-find-the-town-judge.py
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        # We know there canbe one jude atmax
        # i.e. atmax one will have 
        # incoming edges = n - 1 and also outgoing edges = 0
        # if noone has then return -1

        delta = defaultdict(int) # delta = (incoming - outgoing)

        for(src, dst) in trust:
            delta[dst] += 1 # incoming edge
            delta[src] -= 1 # outgoing edge

        for i in range(1, n + 1):
            if delta[i] == n - 1:
                return i

        return -1

File: 1039-find-the-town-judge/1039-find-the-town-judge.cpp
class Solution
{
public:
    int findJudge(int n, vector<vector<int>>& trust)
    {
        vector<bool> possible(n, true);
        map<int, int> mp;
        for(auto pair : trust) possible[pair[0] - 1] = false;
        for(int i = 0; i < n; i++) if(possible[i] == true) mp[i + 1] = n - 1;
        for(auto pair : trust)
        {
            mp[pair[1]]--;
            if(mp[pair[1]] == 0) return pair[1];
        }
        for(auto ele : mp) if(ele.second == 0) return ele.first;
        return -1;
    }
};